text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Mixed Shadow Mode
+status: DRAFTED
+created_at: 2021-05-07
+updated_at: YYYY-MM-DD
+pr: https://github.com/salesforce/lwc-rfcs/pull/49
+---
+
+# Mixed Shadow Mode
+
+## Introduction
+
+LWC implements a collection of polyfills that implement Shadow DOM features. These polyfills are
+collectively referred to as "synthetic Shadow DOM" and are published to npm under the
+`@lwc/synthetic-shadow` package. Synthetic shadow DOM provides a consistent experience for LWC web
+components across the Salesforce-supported browser matrix.
+
+Native Shadow DOM support has greatly increased since the start of the LWC project but applications
+continue to rely on the synthetic shadow polyfills to maintain backwards-compatibility. The removal
+of these polyfills can result in broken experiences due to dependencies on certain "escape hatches"
+that are not possible in a native Shadow DOM environment (e.g., instrumentation, styling, etc, which
+are discussed below).
+
+Applications that wish to use LWC components in a native shadow context simply need to omit the
+inclusion of the `@lwc/synthetic-shadow` polyfill; however, this is not a realistic option for
+large applications that cannot afford a rewrite.
+
+This proposal opens up an incremental migration path for applications moving to native Shadow DOM by
+allowing the usage of both native and synthetic Shadow DOM in the same document.
+
+## Detailed design
+
+Components will opt-in to this new feature and the default mode will be synthetic mode. From the
+component's perspective, there are two choices of shadow semantics: synthetic mode or native
+mode. The concept of a mixed mode should be transparent and the framework should guarantee this
+through integration testing, which is further discussed below.
+
+A component can signal that it prefers native Shadow DOM by setting a static property called
+`preferNativeShadow` to `true`.

```

Comment:
I'd just like to point out, since it came up in the light DOM discussion, that we are currently defaulting this to `false` but someday (presumably) it will default to `true`. So we can't really avoid the "booleans should default to `false`" problem here.

Reply:
I think you mean to say that many components will set this property to `true` but I don't think we can ever default it to `true`...at least without versioning. There are thousands of LWC components in existence today and we can't expect all of them to support mixed mode. My guess is that applications that are ready for a change like defaulting this to `true` will simply stop including the polyfill.

Reply:
That's a good point. I guess I mean that, at some point in the future, LEX will be 100% native shadow DOM, correct? Which means that effectively the default will be `true` from the component author's perspective?

Reply:
Yeah, I see what you mean. I guess that open source applications would just exclude the synthetic shadow polyfill, but we would probably need to do something different for platform as the majority of components start using native. We could probably do something smart in the platform compiler.